api/envoy/config/listener/v3/listener.proto

```diff
@@ -120,6 +121,66 @@ message Listener {
   // :ref:`FAQ entry <faq_how_to_setup_sni>`.
   repeated FilterChain filter_chains = 3;
 
+  // Unified matcher resolving the filter chain name from the network properties. This matcher is used as a replacement
+  // for the per-filter chain match condition
+  // `filter_chain_match <envoy_v3_api_msg_config.listener.v3.FilterChain.filter_chain_match>`.
+  // If specified, all :ref:`filter_chains <envoy_v3_api_field_config.listener.v3.Listener.filter_chains>` must
+  // have non-empty and unique :ref:`name <envoy_v3_api_field_config.listener.v3.FilterChain.name>` fields and omit
+  // `filter_chain_match <envoy_v3_api_msg_config.listener.v3.FilterChain.filter_chain_match>` field.
+  //
+  // Example 1: The following matcher selects three filter chains as follows:
+  //
+  // * if the destination port is 80, then the filter chain "http" is selected;
+  // * if the destination port is 443 and the source IP is in the range 192.0.0.0/2, then the filter chain "internal" is selected;
+  // * otherwise, if the destination port is 443, then the filter chain "https" is selected;
+  // * otherwise, the default filter chain is selected (or the connection is rejected without the default filter chain).
+  //
+  // .. code-block:: yaml
+  //
+  //  filter_chain_matcher:
+  //    matcher_tree:
+  //      input:
+  //        typed_config:
+  //          "@type": type.googleapis.com/envoy.extensions.matching.common_inputs.network.v3.DestinationPortInput
+  //      exact_match_map:
+  //        map:
+  //          "80":
+  //            action:
+  //              name: http
+  //          "443":
+  //            matcher:
+  //              matcher_tree:
+  //                input:
+  //                  typed_config:
+  //                    "@type": type.googleapis.com/envoy.extensions.matching.common_inputs.network.v3.SourceIPInput
+  //                custom_match:
+  //                  typed_config:
+  //                    "@type": type.googleapis.com/xds.type.matcher.v3.IPMatcher
+  //                    range_matchers:
+  //                    - ranges:
+  //                      - address_prefix: 192.0.0.0
+  //                        prefix_len: 2
+  //                      on_match:
+  //                        action:
+  //                          name: internal
+  //                    - ranges:
+  //                      - address_prefix: 0.0.0.0
+  //                      on_match:
+  //                        action:
+  //                          name: https
+  //
+  //
+  // .. note::
+  //
+  //  Once matched, each connection is permanently bound to its filter chain.
+  //  If the matcher changes but the filter chain remains the same, the
+  //  connections bound to the filter chain are not drained. If, however, the
+  //  filter chain is removed or structurally modified, then the drain for its
+  //  connections is initiated.
+  //
+  // [#not-implemented-hide:]
```

Comment:
I see this configuration as being one of the ones that will be very difficult for users to actually piece together without examples. Can we make sure that somehow we wire up the extension docs system here so that this somehow lists out all supported match inputs and relevant typed configs, etc.? We need to guide users much more specifically on how to use this new field.

Reply:
Agree, we need more examples. I can add some but I'd really need https://github.com/envoyproxy/envoy/pull/19493 merged first for protos to become available. There is an issue with cncf/xds protos not being documented.